### PR TITLE
Hsa1/booktrak add buy listings page

### PR DESCRIPTION
**Pull request by harryalbert**
updated book listings component to also work as the buy/sell listings page

Demo:
https://github.com/WilliamsStudentsOnline/wso-react/assets/65182701/e87e71be-9f46-45d9-ae3c-da0726ade07d

### CODE REVIEW

File: src/components/views/Booktrak/BooktrakListings.tsx [hunk 1]
```diff
@@ -4,98 +4,158 @@ import React, { useState, useEffect } from "react";
...
+import "../../stylesheets/Booktrak.css";
+import { ModelsBookListingWithTitle } from "./BooktrakUtils";
 
-const BooktrakListings = ({ book }: { book: ModelsBook }) => {
+type ServerError = {
+  errorCode: number;
+  message: string;
+};
+
+function isServerError(error: unknown): error is ServerError {
```
Comment (AndrewMuh) [inline src/components/views/Booktrak/BooktrakListings.tsx:24]: Move this to `lib/general.js`. Also rewrite it as `export const isServerError` instead of function.

File: src/components/views/Booktrak/BooktrakUtils.tsx [hunk 2]
```diff
@@ -1,5 +1,9 @@
 import { ModelsBookListing } from "wso-api-client/lib/services/types";
 
+export interface ModelsBookListingWithTitle extends ModelsBookListing {
```
Comment (AndrewMuh) [inline src/components/views/Booktrak/BooktrakUtils.tsx:3]: Remove and use ModelsBookListing.ModelsBook.title

File: src/components/views/Booktrak/BooktrakListings.tsx [hunk 3]
```diff
@@ -4,98 +4,158 @@ import React, { useState, useEffect } from "react";
...
           }
-        );
-        updateBuyListings(buyListingsResponse.data ?? []);
-        updateSellListings(sellListingsResponse.data ?? []);
+          updateListings(newListings ?? []);
+        }
       } catch (error) {
-        // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        navigateTo("/404", { replace: true });
+        if (isServerError(error)) {
```
Comment (AndrewMuh) [inline src/components/views/Booktrak/BooktrakListings.tsx:104]: Should probably still navigate to error page if there is a server error

File: src/components/views/Booktrak/BooktrakListings.tsx [hunk 4]
```diff
@@ -4,98 +4,158 @@ import React, { useState, useEffect } from "react";
...
+  const [listings, updateListings] = useState<ModelsBookListingWithTitle[]>([]);
+  const [buyListings, updateBuyListings] = useState<
+    ModelsBookListingWithTitle[]
+  >([]);
+  const [sellListings, updateSellListings] = useState<
+    ModelsBookListingWithTitle[]
+  >([]);
+
+  const [currentPage, updateCurrentPage] = useState(0);
+  const maxListingsPerPage = 20;
```
Comment (AndrewMuh) [inline src/components/views/Booktrak/BooktrakListings.tsx:54]: Use the backend pagination to paginate. You can use the `offset` field in `listBookListings`

**Review by AndrewMuh — CHANGES_REQUESTED** [review]
TL;DR use new backend API functionality to preload book and user

File: src/components/views/Booktrak/BooktrakListings.tsx [hunk 5]
```diff
@@ -4,98 +4,158 @@ import React, { useState, useEffect } from "react";
...
+    "errorCode" in error &&
+    "message" in error
+  );
+}
+
+const ListingTypeEnum = ModelsBookListing.ListingTypeEnum;
+const BooktrakListings = ({
+  book,
+  onlyShowBuyListings,
+  displayBuyAndSell,
```
[the remark this answers is not in the document]
Comment (harryalbert) [reply src/components/views/Booktrak/BooktrakListings.tsx:37]: Sounds good. I can't think of a reason you'd ever want to show neither, but I'll just return an empty component in that case

File: src/components/views/Booktrak/BooktrakListings.tsx (hunk 3 above)
Comment (harryalbert) [reply src/components/views/Booktrak/BooktrakListings.tsx:93]: The reason I did this is because the useEffect gets called a few times when the page loads, and for whatever reason the first call gets a 401 unauthorized response

Comment (AndrewMuh) [reply src/components/views/Booktrak/BooktrakListings.tsx]: You should fix the reason this happens. It's because the route isn't wrapped around a RequireScope component [here](https://github.com/WilliamsStudentsOnline/wso-react/blob/506a589bb8f09604e055bf2b0cf6d4736753f7e0/src/components/App.tsx#L182). Also the module should be lazily imported like the rest instead of directly imported [here](https://github.com/WilliamsStudentsOnline/wso-react/blob/506a589bb8f09604e055bf2b0cf6d4736753f7e0/src/components/App.tsx#L27). Make sure to add token level 3 for bookmark [here](https://github.com/WilliamsStudentsOnline/wso-react/blob/506a589bb8f09604e055bf2b0cf6d4736753f7e0/src/router-permissions.js#L14) for this to work as well.

**Review by AndrewMuh — CHANGES_REQUESTED** [review]
Fix route problems

**Review by AndrewMuh — APPROVED** [review]
LGTM
